File: src/client/gochat_client/lib/models/message.dart
enum MessageType {
  text(1),
  image(2),
  video(3);

  final int value;
  const MessageType(this.value);

  static MessageType fromValue(int value) {
    return MessageType.values.firstWhere((e) => e.value == value);
  }
}

enum MessageStatus {
  sending,
  sent,
  delivered,
  read,
  failed,
}

class Message {
  final String msgId;
  final int fromUserId;
  final int toUserId;
  final MessageType msgType;
  final String content;
  final bool isGroup;
  final int? groupId;
  final DateTime createTime;
  MessageStatus status;
  final bool isRevoked;
  final DateTime? revokeTime;

  Message({
    required this.msgId,
    required this.fromUserId,
    required this.toUserId,
    required this.msgType,
    required this.content,
    this.isGroup = false,
    this.groupId,
    required this.createTime,
    this.status = MessageStatus.sent,
    this.isRevoked = false,
    this.revokeTime,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      msgId: json['msgId'] as String,
      fromUserId: json['fromUserId'] as int,
      toUserId: json['toUserId'] as int,
      msgType: MessageType.fromValue(json['msgType'] as int),
      content: json['content'] as String,
      isGroup: json['isGroup'] as bool? ?? false,
      groupId: json['groupId'] as int?,
      createTime: DateTime.parse(json['createTime'] as String),
      isRevoked: json['isRevoked'] as bool? ?? false,
      revokeTime: json['revokeTime'] != null
          ? DateTime.tryParse(json['revokeTime'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'msgId': msgId,
      'fromUserId': fromUserId,
      'toUserId': toUserId,
      'msgType': msgType.value,
      'content': content,
      'isGroup': isGroup,
      'groupId': groupId,
      'createTime': createTime.toIso8601String(),
      'isRevoked': isRevoked,
      'revokeTime': revokeTime?.toIso8601String(),
    };
  }

  bool get isMine => false; // 将在使用时根据当前用户判断

  // 判断消息是否可以撤回（2分钟内）
  bool canRecall(int currentUserId) {
    if (isRevoked) return false;
    if (fromUserId != currentUserId) return false;
    final now = DateTime.now();
    final difference = now.difference(createTime);
    return difference.inMinutes < 2;
  }

  // 创建撤回后的消息副本
  Message copyWithRevoked() {
    return Message(
      msgId: msgId,
      fromUserId: fromUserId,
      toUserId: toUserId,
      msgType: msgType,
      content: content,
      isGroup: isGroup,
      groupId: groupId,
      createTime: createTime,
      status: status,
      isRevoked: true,
      revokeTime: DateTime.now(),
    );
  }
}

File: src/server/gochat-server/ent/schema/message.go
package schema

import (
	"time"

	"entgo.io/ent"
	"entgo.io/ent/schema/field"
)

// Message 基础消息表：存储消息元信息（类型、内容、创建时间）
// 具体内容表 TextMessage/ImageMessage/VideoMessage 通过 msgId 关联
type Message struct {
	ent.Schema
}

// Fields of the Message.
func (Message) Fields() []ent.Field {
	return []ent.Field{
		field.String("msgId").NotEmpty().Comment("消息ID,由发送者产生"),
		field.String("msgType").NotEmpty().Comment("消息类型: text/image/video 或数值枚举"),
		field.String("content").NotEmpty().Comment("消息内容(冗余存储,便于快速列表展示)"),
		field.Bool("isRevoked").Default(false).Comment("是否已撤回"),
		field.Time("revokeTime").Optional().Nillable().Comment("撤回时间"),
		field.Time("createTime").Default(time.Now).Comment("创建时间"),
	}
}

// Edges of the Message.
func (Message) Edges() []ent.Edge {
	return nil
}

